compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,220 @@
+        // it again because there might be multiple calls to the same brillig function.
+        let mut function = Function::clone_with_id(*func_id, function);
+
+        // Find the entry block and remove its parameters
+        let entry_block_id = function.entry_block();
+        let entry_block = &mut function.dfg[entry_block_id];
+        let entry_block_parameters = entry_block.take_parameters();
+
+        assert_eq!(arguments.len(), entry_block_parameters.len());
+
+        // Replace the ValueId of parameters with the ValueId of arguments
+        for (parameter_id, argument_id) in entry_block_parameters.iter().zip(arguments) {
+            // Lookup the argument in the current function and insert it in the function copy
+            let new_argument_id = copy_constant_to_function(self, &mut function, *argument_id);
+            function.dfg.set_value_from_id(*parameter_id, new_argument_id);
+        }
+
+        // Try to fully optimize the function. If we can't, we can't inline it's constant value.
+        let Ok(mut function) = optimize(function, inliner_aggressiveness, error_selector_to_type)
+        else {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        };
+
+        let entry_block = &mut function.dfg[entry_block_id];
+
+        // If the entry block has instructions, we can't inline it (we need a terminator)
+        if !entry_block.instructions().is_empty() {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        let terminator = entry_block.take_terminator();
+        let TerminatorInstruction::Return { return_values, call_stack: _ } = terminator else {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        };
+
+        // Sanity check: make sure all returned values are constant
+        if !return_values.iter().all(|value_id| function.dfg.is_constant(*value_id)) {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        // Replace the instruction results with the constant values we got
+        let current_results = self.dfg.instruction_results(instruction_id).to_vec();
+        assert_eq!(return_values.len(), current_results.len());
+
+        for (current_result_id, return_value_id) in current_results.iter().zip(return_values) {
+            let new_return_value_id = copy_constant_to_function(&function, self, return_value_id);
+            self.dfg.set_value_from_id(*current_result_id, new_return_value_id);
+        }
+
+        true
+    }
+}
+
+/// Copies a constant from one function to another.
+/// Though it might seem we can just take a value out of `from_function` and call `make_value` on `to_function`,
+/// if the constant is an array the values will still keep pointing to `from_function`. So, this function
+/// recursively copies the array values too.
+fn copy_constant_to_function(
```

Comment:
Just curious: why is `optimize_const_brillig_call` defined on `Function`, but this one, which could also be `self.copy_constant_from_function(&mut self, from_function: &Function, constant_id: ValueId)`, defined as a separate function?  I thought maybe you don't want it to pollute the `Function` namespace in other places, but if the other method is there, rather than on some kind of `Context` like in some of the other packages, why not this? It also looks like a method that could find its uses elsewhere.

Reply:
Oh, good point! Actually, I had this defined in `Function` originally, but then I needed to copy constants from the second function into the first one so I thought "I'll need to make this a free function" but I forgot I could call it on the second function 😓 

I'll change it, I think it's clearer as a method.

Reply:
Done! I left it in this same file. If later this functionality is generally needed we could move it to the main `Function` impl.